### PR TITLE
[clang] Track function template instantiation from definition

### LINKED ISSUE #55509: Clang crashes in clang::Sema::BuildExpressionFromIntegralTemplateArgument with c++2b
The following code makes clang to crash. See [at godbolt](https://godbolt.org/z/4q6co9Gae).

```cpp
template<int N>
struct A {
    template<class C>
    friend auto cica(const A<N-1>&, C) {
        return N-1;
    }
};

template<>
struct A<0> {
    template<class C>
    friend auto cica(const A<0>&, C);
};

template <int N, class AT, class = decltype(cica(AT{}, nullptr))>
static constexpr bool MakeCica(int);

template <int N, class AT>
static constexpr bool MakeCica(short, A<N+1> = {});

template <int N, class AT = A<N>, class Val = decltype(MakeCica<N, AT>(0))>
static constexpr bool has_cica = Val{};

constexpr bool cica2 = has_cica<0> || has_cica<0>;
```

### PR DESCRIPTION
**Pull request by mizvekov**
This fixes instantiation of definition for friend function templates,
when the declaration found and the one containing the definition
have different template contexts.

In these cases, the the function declaration corresponding to the
definition is not available; it may not even be instantiated at all.

So this patch adds a bit which tracks which function template
declaration was instantiated from the member template.
It's used to find which primary template serves as a context
for the purpose of obtaining the template arguments needed
to instantiate the definition.

Fixes #55509

### CODE REVIEW

File: clang/lib/Sema/SemaTemplateDeduction.cpp [hunk 1]
```diff
@@ -3949,28 +3949,15 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
       TemplateArgumentList::CreateCopy(Context, CanonicalBuilder);
   Info.reset(SugaredDeducedArgumentList, CanonicalDeducedArgumentList);
 
+  FunctionTemplate = FunctionTemplate->getMostRecentDecl();
+
```
Comment (zyn0217) [inline clang/lib/Sema/SemaTemplateDeduction.cpp:3953]: So this is where the magic happens:

To help us get a bigger picture of the patch, let's take an example from your test case,

```cpp
namespace t1 {
  template<int N> struct A {
    template<class C> friend auto cica(const A<N-1>&, C) {
      return N;
    }
  };

  template<> struct A<0> {
    template<class C> friend auto cica(const A<0>&, C);
  };

  void test() {

    (void)A<1>{};
    cica(A<0>{}, 0);
  }
} // namespace t1
```

Previously, we would crash in `BuildExpressionFromIntegralTemplateArgument()` because we're trying to substitute the NTTP `N` in the return expression with a template argument `<int>`. That is due to an incorrect MLTAL provided by `getTemplateInstantiationArgs()`, when instantiating the function definition of `cica` for the purpose of return type deduction.

The MLTAL is missing the outer `N` because the collected template arguments are from a friend function whose parent is an explicit template specialization, `template <> struct A<0>`. And we don't do anything else for an explicit specialization, so we're left with only one level of template arguments, which is just `<int>`.

The approach here did the trick that, instead of instantiating the function from the explicit specialization, it finds the declaration that is created by the last implicit instantiation of the corresponding explicit specialization, namely the specialization for `A<N>`.

I presume changes elsewhere are tweaking to adapt the new behavior here, right?

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 2]
```diff
@@ -2146,9 +2147,15 @@ Decl *TemplateDeclInstantiator::VisitFunctionDecl(
...
-      return SpecFunc;
+    if (SpecFunc) {
+      if (!SpecFunc->isTemplateInstantiation())
+        return SpecFunc;
+
+      for (auto *Redecl : SpecFunc->redecls())
+        if (Redecl->getPrimaryTemplate() == FunctionTemplate)
+          return Redecl;
+    }
+    PrevFunc = SpecFunc;
```
Comment (zyn0217) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:2158]: Let me rephrase the purpose here:

So now clang instantiates a new function declaration as long as the previous instantiation doesn't exist / the previous instantiation has a distinct primary template declaration, right?

I feel I struggle to understand the benefit here, though it might indeed have fixed/improved something. Can you explain it more?

File: clang/test/SemaTemplate/GH55509.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c++26 %s
```
Comment (zyn0217) [inline clang/test/SemaTemplate/GH55509.cpp:1]: nit: we might want to move it to an existing file (so we don't have tests scattered around), but it's up to you if you feel the test is long enough to have a new file.

**Review by zyn0217 — COMMENTED** [review]
Thanks!

File: clang/lib/Sema/SemaTemplateDeduction.cpp (hunk 1 above)
Comment (mizvekov) [reply clang/lib/Sema/SemaTemplateDeduction.cpp:3953]: > The MLTAL is missing the outer N because the collected template arguments are from a friend function whose parent is an explicit template specialization, template <> struct A<0>. And we don't do anything else for an explicit specialization, so we're left with only one level of template arguments, which is just <int>.

So far correct, yes.

> The approach here did the trick that, instead of instantiating the function from the explicit specialization, it finds the declaration that is created by the last implicit instantiation of the corresponding explicit specialization, namely the specialization for A<N>.

Well not exactly. The problem here is that in this call:
```C++
cica(A<0>{}, 0);
```
Due to ADL, we find the declaration of `cica` in the explicit specialization `A<0>`, instead of the more recent one created in the explicit instantiation `A<1>`.

We end up crashing because starting from the instantiation of the second declaration of cica, we find the pattern for the first declaration, which has incorrect template context as you remarked above.

> I presume changes elsewhere are tweaking to adapt the new behavior here, right?

For the most part yes, but the more interesting changes in `SemaTemplateInstantiateDecl.cpp` are dealing with a more complicated case. What happens in:

```
  cica(A<0>{}, 0);
  (void)A<1>{};
  cica(A<0>{}, 0);
```

Ie: when we instantiate `A<1>`, we have already implicitly instantiated the cica declaration for `A<0>`, and on the second call to `cica`, that is the implicit instantiation that we are going to find, and again we are going to find the wrong pattern for it and crash.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp (hunk 2 above)
Comment (mizvekov) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:2158]: This fixes the case:
```C++
  cica(A<0>{}, 0);
  (void)A<1>{};
  cica(A<0>{}, 0);
```

As explained in the previous comment.

File: clang/lib/Sema/SemaTemplateDeduction.cpp (hunk 1 above)
Comment (zyn0217) [reply clang/lib/Sema/SemaTemplateDeduction.cpp:3953]: > Ie: when we instantiate A<1>, we have already implicitly instantiated the cica declaration for A<0>, and on the second call to cica, that is the implicit instantiation that we are going to find, and again we are going to find the wrong pattern for it and crash.

Oh I see! Thanks for the detailed explanation, the logic of reusing previous instantiations is sometimes really a footgun and this is indeed a clever fix.

**Review by zyn0217 — APPROVED** [review]
Everything makes sense to me now, thank you!

**Review by cor3ntin — APPROVED** [review]
LGTM once @zyn0217 outstanding comments are resolved.

File: clang/test/SemaTemplate/GH55509.cpp (hunk 3 above)
Comment (mizvekov) [reply clang/test/SemaTemplate/GH55509.cpp:1]: Yeah I think this is big enough to be on its own file.

**Comment by mizvekov** [comment]
I pushed a new implementation for this fix.

I have updated the implementation of InstantiateFunctionDefinition so that it doesn't depend anymore on having the exact function declaration instantiation which has the same template context as the definition.

This avoids the more expensive aspect of the previous approach.
While keeping track of separate function declaration instantiations per primary template redeclaration increases source representation accuracy, It would be better to, in the future, go that way across that board for all templates, not just function templates, so that the impact can be better represented.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 4]
```diff
@@ -5185,9 +5189,24 @@ void Sema::InstantiateFunctionDefinition(SourceLocation PointOfInstantiation,
...
+    std::optional<ArrayRef<TemplateArgument>> Innermost;
+    if (auto *Primary = Function->getPrimaryTemplate();
+        Primary &&
+        !isGenericLambdaCallOperatorOrStaticInvokerSpecialization(Function)) {
+      ND = *llvm::find_if(Primary->redecls(),
+                          [](const RedeclarableTemplateDecl *RTD) {
+                            auto *FTD = cast<FunctionTemplateDecl>(RTD);
+                            return FTD->isInstantiatedFromDefinition() ||
+                                   FTD->isThisDeclarationADefinition();
+                          });
```
Comment (zyn0217) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5202]: We'd better have an assertion here, ND shouldn't be a nullptr.

**Review by zyn0217 — COMMENTED** [review]
Thanks for taking another look at it.

My main concern is that changes in `getTemplateInstantiationArgs()` would conflict with @sdkrystian's recent work, so if these changes are necessary, I suggest we merge this one after the refactoring patch lands, WDYT?

File: clang/lib/Sema/SemaTemplateInstantiate.cpp [hunk 5]
```diff
@@ -492,8 +489,10 @@ MultiLevelTemplateArgumentList Sema::getTemplateInstantiationArgs(
     // has a depth of 0.
     if (const auto *TTP = dyn_cast<TemplateTemplateParmDecl>(CurDecl))
       HandleDefaultTempArgIntoTempTempParam(TTP, Result);
-    CurDecl = Response::UseNextDecl(CurDecl).NextDecl;
-  }
+    CurDecl = DC ? Decl::castFromDeclContext(DC)
+                 : Response::UseNextDecl(CurDecl).NextDecl;
```
[the remark this answers is not in the document]
Comment (mizvekov) [reply clang/lib/Sema/SemaTemplateInstantiate.cpp:493]: In this case, we don't even have the real function declaration that we want to start with, so I am afraid this is too much of an special case  to handle within `getTemplateInstantiationArgs()`.

I think the best way to handle this one is to have one overload of `getTemplateInstantiationArgs` which takes an initial MLTAL, and just adds more template arguments on top of it.

Yeah, I am planning on holding this off a little bit until the refactoring lands.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp (hunk 4 above)
Comment (mizvekov) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5202]: We are finding one declaration within all the redeclarations of the primary template.
The list of redeclarations can't have a null element in it.

Otherwise, if you meant a null iterator, we shouldn't get here without a definition, as we should have diagnosed that before.
We could perhaps add a null check for that. Is that what you meant?

Comment (zyn0217) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5202]: Oh yeah, I think I was saying the iterator shouldn't be null.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 6]
```diff
@@ -1994,8 +1995,10 @@ TemplateDeclInstantiator::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
   // Link the instantiation back to the pattern *unless* this is a
   // non-definition friend declaration.
   if (!InstTemplate->getInstantiatedFromMemberTemplate() &&
-      !(isFriend && !D->getTemplatedDecl()->isThisDeclarationADefinition()))
+      !(isFriend && !D->getTemplatedDecl()->isThisDeclarationADefinition())) {
+    InstTemplate->setInstantiatedFromDefinition();
```
[the remark this answers is not in the document]
Comment (mizvekov) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:1999]: That's true, but the only problematic cases are the friends.

We could make the flag more accurate, as named, for non-friends, but then we don't have a motivation, and we would make searching for a declaration which is compatible with the definition slightly more expensive for those.

Another option is to rename the flag to something like `InstantiatedFromDefinitionForFriend`.

[the remark this answers is not in the document]
Comment (zyn0217) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:1999]: > That's true, but the only problematic cases are the friends.

I was thinking if we can move it to a separate if branch then, for example it might not be conditioned on `isFriend`

```cpp
if (D->getTemplatedDecl()->isThisDeclarationADefinition())
  InstTemplate->setInstantiatedFromDefinition();
```

(I presume a new if won't impact anything other than friends, as you said the friends are the only problematic ones.)

**Comment by mizvekov** [comment]
Now this is rebased on top of the `getTemplateInstantiationArgs` refactoring.

File: clang/include/clang/AST/DeclTemplate.h [hunk 7]
```diff
@@ -1008,6 +1008,15 @@ class FunctionTemplateDecl : public RedeclarableTemplateDecl {
...
   }
 
+  bool isCompatibleWithDefinition() const {
+    return getTemplatedDecl()->isInstantiatedFromMemberTemplate() ||
+           isThisDeclarationADefinition();
+  }
+  void setInstantiatedFromMemberTemplate(FunctionTemplateDecl *D) {
+    getTemplatedDecl()->setInstantiatedFromMemberTemplate();
+    RedeclarableTemplateDecl::setInstantiatedFromMemberTemplate(D);
+  }
```
Comment (zyn0217) [inline clang/include/clang/AST/DeclTemplate.h:1018]: So the flag on FunctionDecl now appears to be just a shortcut of getDescribedTemplate()->getInstantiatedFromMemberTemplate(), or am I missing something else?

Comment (mizvekov) [reply clang/include/clang/AST/DeclTemplate.h:1018]: getInstantiatedFromMemberTemplate() uses the common area, which means whenever you set it for one declaration, it's set the same for all its redeclarations.
But the flag does not use the common area, it's a separate flag per redeclaration.

Comment (zyn0217) [reply clang/include/clang/AST/DeclTemplate.h:1018]: Get it, thanks!
